### DISCUSSION TITLE
How to get the line number where link reference definition is defined?

### DISCUSSION THREAD

**Discussion by LastDragon-ru** [thephpleague/commonmark#1036] [Q&A]
I want to update link references in markdown, until #419 is done, the only way is to replace it in the text. But I cannot find a way how to determine the line where reference is defined :(

* `ReferenceInterface` doesn't hold this information at all
* Custom block parser seems will break [all references](https://github.com/thephpleague/commonmark/blob/0026475f5c9a104410ae824cb5a4d63fa3bdb1df/src/Parser/MarkdownParser.php#L254-L256)

For example, for this document the line is `3`.

```markdown
Paragraph with [link][target].

[target]: https://example.com
```

**Comment by camerongreen** [comment]
I had an issue with reference links which were part of the Google Docs markdown export. The HTML output kept them as seperate elements, leaving the original markdown reference intact, and converting the link to a standardanchor  href.  I'd like to know if I've missed something in the documentation about how to handle this.   

What I've ended up doing was creating a wrapper function to convert the reference links to individual anchors before processing the markdown.   Might need some testing:

```php
  /**
   * Process a markdown string with link references.
   *
   * @param string $markdown
   *   The markdown string to process.
   *
   * @return string
   *   A string containing the processed HTML.
   */
  public function processMarkdown(string $markdown): string {
    // Find all reference-style image definitions (e.g., [image1]: <data:...>).
    preg_match_all('/^\[([^\]]+)\]:\s*<([^>]+)>/m', $markdown, $matches, PREG_SET_ORDER);

    $image_references = [];
    foreach ($matches as $match) {
      $id = $match[1];
      $url = $match[2];
      $image_references[$id] = $url;

      // Remove the definition line from the markdown.
      $markdown = str_replace($match[0], '', $markdown);
    }

    // Replace reference-style image links with inline-style links.
    // e.g., ![Alt text][image1] becomes ![Alt text](data:...).
    $markdown = preg_replace_callback(
      '/!\[([^\]]*)\]\[([^\]]+)\]/',
      function ($link_match) use ($image_references) {
        $alt_text = $link_match[1];
        $id = $link_match[2];
        return $image_references[$id] ? "![{$alt_text}]({$image_references[$id]})" : $link_match[0];
      },
      $markdown
    );

    $converter = new CommonMarkConverter();
    return $converter->convert($markdown)->getContent();
}
```
